src/lib/Embed.hs

```diff
@@ -783,21 +756,19 @@ intToIndexE (VariantTy types) i = do
     -- TODO: is there a better way than generating a case statement? Select
     -- doesn't work because variants aren't primitive types
     go prev ((label, repeatNum), ty, offset) = do
+      shifted <- isub i offset
+      -- TODO: This might run intToIndex on negative indices. Fix this!
+      index   <- intToIndexE ty shifted
       beforeThis <- ilt i offset
-      boolCase beforeThis (return prev) $ do
-        shifted <- isub i offset
-        index <- intToIndexE ty shifted
-        return $ Variant types label repeatNum index
-    ((l0, 0), ty0, _):zs =
-      zip3 (toList reflect) (toList types) (toList offsets)
+      select beforeThis prev $ Variant types label repeatNum index
```

Comment:
Does this work? I thought I tried something like this and ran into problems because the `Select` op implementation assumed its inputs were basic types, and Variants are not basic types. Maybe something else in this PR fixes that too? (edit: I think I understand now, see below)